### PR TITLE
spike: Better address management (SC-15104)

### PR DESCRIPTION
**Pull request by JGcarv**

### CODE REVIEW

File: src/generate-md.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+const capitalize = (s) => s.charAt(0).toUpperCase() + s.slice(1);
+
+
+registry.chains.forEach(chain => {
+
+    let content = [];
+    
+    for (const [key, value] of Object.entries(chain.contracts)) {  
+        
+        if (key == "pools") {
```
Comment (vbidin) [inline src/generate-md.js:25]: Could replace the `if` statements with a `switch`, and throw errors if the key is not recognized. Same for the Solidity generator script.

File: src/generate-md.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+} from './utils.js';
+
+
+const registry = await importJson("../address-registry.json");
+
+const section = (title, depth = 1) => `${'#'.repeat(depth)} ${title}`;
+
+const tableHeader = `| Contract | Address |\n| --- | --- |`;
+
+const declaration = (key, value) => `| ${key} | [${value}](https://etherscan.io/address/${value}) |`;
```
Comment (vbidin) [inline src/generate-md.js:14]: Should use `basescan.org` for Base contracts.

File: contracts/MapleAddressRegistryBASEL2.sol [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+
+    /**************************************************************************************************************************************/
+    /*** Assets                                                                                                                         ***/
+    /**************************************************************************************************************************************/
+
+    address constant public usdc = 0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913;
+    address constant public weth = 0x4200000000000000000000000000000000000006;
+
+}
```
Comment (vbidin) [inline contracts/MapleAddressRegistryBASEL2.sol:103]: Add newlines to the end of the generation scripts.

File: contracts/MapleAddressRegistryBASEL2.sol [hunk 4]
```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: BUSL-1.1 
+pragma solidity ^0.8.7; 
+
+// WARNING: File generated automatically, do not edit manually
+contract MapleAddressRegistryBASEL2 {
+    /**************************************************************************************************************************************/
```
Comment (vbidin) [inline contracts/MapleAddressRegistryBASEL2.sol:6]: Missing a newline here.

File: address-registry.json [hunk 5]
```diff
@@ -0,0 +1,505 @@
+{
+    "chains": [
+        {
+            "name": "ETH",
+            "contracts": {
+                "actors": [
+                    {
+                        "name": "governor",
```
Comment (vbidin) [inline address-registry.json:8]: Noticed some contracts don't have labels, should we include them so the `README` is more readable?

**Review by vbidin — CHANGES_REQUESTED** [review]

File: address-registry.json (hunk 5 above)
Comment (JGcarv) [reply address-registry.json:8]: You mean the protocol-registry md file?

Comment (vbidin) [reply address-registry.json:8]: Yeah

Comment (JGcarv) [reply address-registry.json:8]: I think that maybe the proxy would make sense, although I don't think it's super important. Honestly, it's a bit cumbersome to handle the labels on generation, but if you both think it's worth it, I can do it.

File: contracts/MapleAddressRegistryBASEL2.sol [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    address constant public globals               = 0x7f3C3636208A18c7941BF051807db56864061465;
+    address constant public treasury              = 0x3A5ab01B4E142F41f54a861c984D19C866Ec1736;
+    address constant public poolDeployerV2        = 0x02A609c8Cbb387E7882836C534bcd738E0E48FbF;
+    address constant public poolDeployerV3        = 0x12fB5dbBDB06ab973f047cC46D6bB33ba4d03b96;
+    address constant public feeManager            = 0xc4D7807a5571887F7b50182d072f4fF4654099b0;
+    address constant public fixedTermRefinancerV2 = 0x510276F01C993De918fa04AFFCd64782FE495846;
+    address constant public openTermRefinancer    = 0xb3138B273884696befc074b48923614fb77c2e10;
+    address constant public poolPermissionManager = 0xBe10aDcE8B6E3E02Db384E7FaDA5395DD113D8b3;
```
Comment (0xfarhaan) [inline contracts/MapleAddressRegistryBASEL2.sol:26]: missing protocol health checker

File: MapleAddressRegistryBASEL2.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
+# Actors
+| Contract | Address |
+| --- | --- |
+| governor | [ 0xD9481a47DFe8E5C35Eb809df25cdB5Bc2E24A3bE](https://basescan.io/address/ 0xD9481a47DFe8E5C35Eb809df25cdB5Bc2E24A3bE) |
```
Comment (0xfarhaan) [inline MapleAddressRegistryBASEL2.md:4]: formatting isn't correct can see when you preview the MD file

File: src/generate-md.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
+import { writeFile } from 'fs';
+
+import {
+    importJson,
```
Comment (0xfarhaan) [inline src/generate-md.js:4]: not sure this needs a new line

File: src/generate-solidity.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
+import { writeFile } from 'fs';
+
+import {
+    importJson,
```
Comment (0xfarhaan) [inline src/generate-solidity.js:4]: same

File: src/utils.js [hunk 10]
```diff
@@ -0,0 +1,3 @@
+import { readFile, writeFile } from 'fs/promises';
```
Comment (0xfarhaan) [inline src/utils.js:1]: write file import not used

**Review by 0xfarhaan — COMMENTED** [review]
Why do we have a pool.md file? We should have one file per chain for both .md and .sol files 

Would recommend checking both current sets of .md and .sol files to make sure no addresses are being missed

**Review by 0xfarhaan — DISMISSED** [review]
Good stuff ser!

**Review by 0xfarhaan — APPROVED** [review]

**Pull request merged**